### PR TITLE
Perf fixes

### PR DESCRIPTION
**Pull request by jeremydstone**
This removes a duplicate database dip for user_levels for a script. This reduces a database call in ScriptLevelsController#show which is currently our perf hotspot. There were previously two methods that were getting called in the same page load that independently dipped to the database to get user_levels for a stage. Now we cache the results in an instance variable in the User object. This is a per-request cache that only lasts the request (page load) lifetime.

### CODE REVIEW

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -502,8 +511,8 @@ def user_progress_by_stage(stage)
   end
 
   def user_level_for(script_level, level)
-    user_levels.find_by(script_id: script_level.script_id,
-                        level_id: level.id)
+    user_levels_for_script = cached_user_levels_by_script(script_level.script_id)
+    user_levels_for_script[level.id]
```
Comment (ashercodeorg) [inline dashboard/app/models/user.rb]: This method now fetches an entire script of `user_levels`, rather than one. This seems like a regression...

File: dashboard/app/views/levels/_admin.html.haml [hunk 2]
```diff
@@ -36,7 +36,7 @@
         %li (Cannot edit)
 
     %br/
-    This level is in #{@level.script_levels.count} scripts:
+    This level is in #{@level.script_levels.to_a.count} scripts:
```
Comment (ashercodeorg) [inline dashboard/app/views/levels/_admin.html.haml:39]: This is presumably unrelated (preventing `count` from causing a DB dip), so could (should) be a separate PR.

File: dashboard/app/models/user.rb (hunk 1 above)
Comment (ashercodeorg) [inline dashboard/app/models/user.rb]: In my world view, `cached_user_levels_by_script` should be `cache_find_user_level(level_id, script_id = nil, fetch_entire_script:)` or some such, where we preemptively populate the cache with all user_levels in `script_id` if `fetch_entire_script`.

File: dashboard/app/views/levels/_admin.html.haml (hunk 2 above)
Comment (ashercodeorg) [inline dashboard/app/views/levels/_admin.html.haml:39]: Aside, I don't think this prevents a DB dip. The dip isn't caused by `count` but rather by `script_levels`, according to `dashboard-console` locally.

Comment (ashercodeorg) [inline dashboard/app/views/levels/_admin.html.haml:39]: Note also that this only called for a handful of users (`LEVELBUILDERS`) on production.

Comment (jeremydstone) [reply dashboard/app/views/levels/_admin.html.haml:39]: Yes, could have been a separate PR, will consider that in the future.

Comment (jeremydstone) [reply dashboard/app/views/levels/_admin.html.haml:39]: This does appear to in fact prevent a DB dip. script_levels are already loaded and cached. The ActiveRecord count() method appears to trigger a database call -- SELECT count(*) from script_levels where...  even though the user_levels are loaded. to_a makes this an array count method of already loaded levels, not the ActiveRecord count. Adding the to_a does remove the database dip, clearly shown before vs after in log output and rack-mini-profiler output. This extra dip shows up now in development mode (not just levelbuilder) because the "Extra Links" section is visible in development mode. Agree this does not impact production. But this is noise that everyone doing perf debugging in the future would have to understand and ignore while doing perf analysis on their local system. This is a future perf debugging tax I am removing.

Comment (ashercodeorg) [reply dashboard/app/views/levels/_admin.html.haml:39]: FWIW, I'm picky about keeping PRs separate. Doing so makes understanding history easier, makes reviewing easier, makes reverting easier, makes everything easier.

Comment (ashercodeorg) [reply dashboard/app/views/levels/_admin.html.haml:39]: You sure there isn't some rails caching going on? In `dashboard-console`, I'm seeing the DB get hit with `@level.script_levels` or `@levels.script_levels.to_a.count`, as long as it is the first usage.

```
irb(main):001:0> ActiveRecord::Base.logger = Logger.new(STDOUT)
=> #<Logger:0x007fec4bf5af98 @progname=nil, @level=0, @default_formatter=#<Logger::Formatter:0x007fec4bf5aed0 @datetime_format=nil>, @formatter=nil, @logdev=#<Logger::LogDevice:0x007fec4bf5ad68 @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mutex=#<Logger::LogDevice::LogDeviceMutex:0x007fec4bf5acc8 @mon_owner=nil, @mon_count=0, @mon_mutex=#<Mutex:0x007fec4bf5abd8>>>>
irb(main):002:0> ActiveRecord::Base.connection_pool.clear_reloadable_connections!
=> nil
irb(main):003:0> @level = Script.get_from_cache(1).script_levels.first.level
=> #<Blockly id: 456808996, game_id: 11, name: "blockly", created_at: "2015-10-16 22:36:26", updated_at: "2015-10-16 22:36:26", level_num: "u_1_1", ideal_level_source_id: nil, solution_level_source_id: nil, user_id: nil, properties: {}, type: "Blockly", md5: nil, published: false, notes: nil>
irb(main):004:0> @level.script_levels
D, [2016-09-14T13:19:07.062667 #4453] DEBUG -- :   ScriptLevel Load (0.6ms)  SELECT `script_levels`.* FROM `script_levels` INNER JOIN `levels_script_levels` ON `script_levels`.`id` = `levels_script_levels`.`script_level_id` WHERE `levels_script_levels`.`level_id` = 456808996
irb(main):005:0> @level.script_levels.to_a.count
=> 1
```

```
irb(main):001:0> ActiveRecord::Base.logger = Logger.new(STDOUT)
=> #<Logger:0x007fb3d65a0f28 @progname=nil, @level=0, @default_formatter=#<Logger::Formatter:0x007fb3d65a0f00 @datetime_format=nil>, @formatter=nil, @logdev=#<Logger::LogDevice:0x007fb3d65a0e88 @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mutex=#<Logger::LogDevice::LogDeviceMutex:0x007fb3d65a0e60 @mon_owner=nil, @mon_count=0, @mon_mutex=#<Mutex:0x007fb3d65a0de8>>>>
irb(main):002:0> ActiveRecord::Base.connection_pool.clear_reloadable_connections!
=> nil
irb(main):003:0> @level = Script.get_from_cache(1).script_levels.first.level
=> #<Blockly id: 456808996, game_id: 11, name: "blockly", created_at: "2015-10-16 22:36:26", updated_at: "2015-10-16 22:36:26", level_num: "u_1_1", ideal_level_source_id: nil, solution_level_source_id: nil, user_id: nil, properties: {}, type: "Blockly", md5: nil, published: false, notes: nil>
irb(main):004:0> @level.script_levels.to_a.count
D, [2016-09-14T13:16:06.793107 #4424] DEBUG -- :   ScriptLevel Load (0.5ms)  SELECT `script_levels`.* FROM `script_levels` INNER JOIN `levels_script_levels` ON `script_levels`.`id` = `levels_script_levels`.`script_level_id` WHERE `levels_script_levels`.`level_id` = 456808996
=> 1
irb(main):005:0> @level.script_levels
=> #<ActiveRecord::Associations::CollectionProxy [#<ScriptLevel id: 1, level_id: nil, script_id: 1, chapter: 1, created_at: "2015-10-16 22:36:26", updated_at: "2015-10-16 22:36:26", stage_id: 1, position: 1, assessment: nil, properties: nil, named_level: nil>]>
```

On another note, upon further reading, I think the `to_a.count` is actually subideal code. For example, per commentary [here](http://batsov.com/articles/2014/02/17/the-elements-of-style-in-ruby-number-13-length-vs-size-vs-count/), we should be using `size` rather than `count` on the array, noting that this might be a noop per commentary [here](http://stackoverflow.com/questions/4550770/count-size-length-too-many-choices-in-ruby).
